### PR TITLE
Add ndk-context.

### LINKED ISSUE #211: Runtime problems when including multiple versions of ndk-glue
Having multiple versions of ndk-glue in the dependency tree leads to crashes because of uninitialized `NATIVE_ACTIVITY`.

Example: Including ndk-glue 0.5. When entering `main()`, `NATIVE_ACTIVITY` 0.5 gets initialized. Including Oboe crate that uses ndk-glue 0.4.0. When enumerating audio devices, i get:
```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', /Users/ric/.cargo/registry/src/github.com-1ecc6299db9ec823/ndk-glue-0.4.0/src/lib.rs:59:39
```

I tried using `[patch.crates-io]` but that does not seem to work.

This bug might be related to using `resolver = "2"` as required by wgpu.

### PR DESCRIPTION
**Pull request by dvc94ch**
Closes #211

### CODE REVIEW

**Review by rklaehn — APPROVED** [review]

**Review by zarik5 — CHANGES_REQUESTED** [review]

**Comment by zarik5** [comment]
As this PR stands, winit will not be able to replace ndk-glue with ndk-context, but maybe this is intended.

**Comment by dvc94ch** [comment]
> As this PR stands, winit will not be able to replace ndk-glue with ndk-context, but maybe this is intended.

This is indeed intended. winit (and presumably a game engine) is highly tied to the runtime and cargo-apk. I think this should be enough for most crates wanting to add some android support. If there are crates for which this isn't sufficient we can collect the exact reasons why not in an issue and then see if we have to take further steps.

**Comment by dvc94ch** [comment]
Ping @MarijnS95

**Comment by dvc94ch** [comment]
@msiglreith looks like @MarijnS95 is on vaccation or something. what are your thoughts on this PR?

**Review by msiglreith — APPROVED** [review]
API-wise it looks good to me. The crate should be listed in the README and probably also include a section regarding `ndk-context` vs `ndk-glue` somewhere

**Comment by dvc94ch** [comment]
So wondering if we shoud remove the `fn native_activity()` from `ndk-glue`. `winit` is using it to access the `AssetManager`. I think there is probably a way to get an `AssetManager` using only `ndk-context`. Maybe marking it as deprecated would get the message accross to people to stop using it.

**Comment by dvc94ch** [comment]
Yep, we should deprecate it [0] shows how to access it using the jni.

- [0] https://stackoverflow.com/questions/22436259/android-ndk-why-is-aassetmanager-open-returning-null/22436260#22436260

**Comment by MarijnS95** [comment]
> looks like @MarijnS95 is on vac~c~ation or something

Not anymore :wink:. Taking a (still rather short) break from open source was well needed.

---

I thought the idea was to remove all `static`s from `ndk-glue` and shove them into `ndk-context` (with the premise that we never do any breaking changes with breaking updates). It is yet unclear to me if it's right to assume that all points of contention over these duplicate `static` variables originate from purely needing the `vm` and/or `activity`? Looks like "deprecating" `native_activity()` is a bit too harsh?

File: ndk-glue/src/lib.rs
Comment (MarijnS95) [inline ndk-glue/src/lib.rs:58]: Not sure if this is a good idea. If it is, at least set a sensible message here (`#[deprecated = "..."]`).

File: ndk-context/src/lib.rs
Comment (MarijnS95) [inline ndk-context/src/lib.rs:75]: `https://crates.io/crates/ndk-glue` just like we link to the `jni` crate?

File: ndk-context/src/lib.rs
Comment (dvc94ch) [reply ndk-context/src/lib.rs:81]: well, it's unsafe to call after main was called (as it can spawn threads and cause data races etc) also it's unsafe to not be initialized before main. I think `#[doc(hidden)]` as originally proposed is probably the right thing to do.

File: ndk-glue/src/lib.rs
Comment (dvc94ch) [reply ndk-glue/src/lib.rs:58]: will remove

**Comment by dvc94ch** [comment]
Yes, think removing the static `NATIVE_ACTIVITY` is the right thing to do (and complementary to ndk-context).  Although not sure about removing `NATIVE_WINDOW`, `INPUT_QUEUE` and `CONTENT_RECT`.

So in general are you ok with the ndk-context approach to partly solving this problem?

**Comment by dvc94ch** [comment]
So removed the NATIVE_ACTIVITY global as I think this is one we agree on and is worth doing.

**Comment by MarijnS95** [comment]
https://github.com/MarijnS95/android-ndk-rs/compare/anativeactivity-instance @dvc94ch This is my draft idea for removing fallible `static` globals entirely, I can open a PR and we can iterate on it in parallel with your changes.

**Comment by dvc94ch** [comment]
@MarijnS95 that looks good, I think it's an improvement for some use cases. But in general I still think we should try to discourage the use of ndk-glue unless it's absolutely necessary. If you open a PR we can merge it and I'll change jni-audio back to using ndk-context.

File: ndk-context/src/lib.rs
Comment (MarijnS95) [inline ndk-context/src/lib.rs:80]: Fwiw you can decide to panic if `ANDROID_CONTEXT` was not `None` to uphold this guarantee. `ANDROID_CONTEXT.replace(AndroidContext { ... }).is_none()`.

**Comment by MarijnS95** [comment]
> But in general I still think we should try to discourage the use of ndk-glue unless it's absolutely necessary.

I think both PRs are a great step in that direction :wink:. Mine still has some open questions though:
- what if a client application already uses the public `instance` field for its own cases?
- `static` is gone now, but what if two different `ndk-glue` versions with incompatible `GlueInstance` layout try to cast and use the same pointer?

Will need some more thoughts, I think I'll get this merged + released + backported ASAP, then PR the migration to most obvious crates and see where to move next.

File: ndk-context/src/lib.rs
Comment (MarijnS95) [inline ndk-context/src/lib.rs:4]: Just in case, perhaps mention here that having this crate duplicated in the crate graph is dangerous (and we should do everything in our power to never make breaking changes again).

File: ndk-examples/examples/jni_audio.rs
Comment (MarijnS95) [inline ndk-examples/examples/jni_audio.rs:11]:
```suggestion
    let vm = unsafe { jni::JavaVM::from_raw(ctx.vm().cast()) }?;
```

Perhaps?

File: ndk-examples/examples/jni_audio.rs
Comment (MarijnS95) [inline ndk-examples/examples/jni_audio.rs:20]:
```suggestion
            ctx.context().cast(),
```

Should work here too?

File: ndk-glue/src/lib.rs
Comment (MarijnS95) [inline ndk-glue/src/lib.rs:170]: And `.cast()` here. Maybe the `ndk-context` docs too?

File: README.md
Comment (MarijnS95) [inline README.md:11]: Not sure if this is strictly an API, but I can't really come up with a short enough alternative. `Runtime handle storage` but that's weird perhaps.

File: ndk-context/src/lib.rs
Comment (MarijnS95) [reply ndk-context/src/lib.rs:4]: Dangerous in the sense that you'll get annoying errors, same as we get right now and are trying to solve by moving to a crate that's a bit more stable.

File: ndk-context/Cargo.toml
Comment (dvc94ch) [reply ndk-context/Cargo.toml:6]: How about "android handles for accessing the jni"? Essentially that's what it is as you can access any android jni api with these.

Comment (dvc94ch) [reply ndk-context/Cargo.toml:6]: or "handles for accessing android apis"

Comment (MarijnS95) [reply ndk-context/Cargo.toml:6]: Perfect.

File: ndk-context/src/lib.rs
Comment (dvc94ch) [reply ndk-context/src/lib.rs:4]: well, I guess that's up to us to make sure we don't release breaking changes. I think it's simple enough that we can say we nailed it.

File: ndk-context/src/lib.rs
Comment (MarijnS95) [inline ndk-context/src/lib.rs:76]:
```suggestion
/// before `main` is called.
```

**Comment by MarijnS95** [comment]
Looks like GH comment notifications are all over the place. Are we removing `native_activity()`, deprecating it at least, or leaving it to a separate PR entirely?

**Comment by MarijnS95** [comment]
Do we need a changelog for the new crate, and the changes listed in the `ndk-glue` crate? Then I think we are ready for merging and releasing. I'm preparing the necessary changes to backport this to older `ndk-glue` crates which are currently used by at least `winit`.

**Comment by dvc94ch** [comment]
> Looks like GH comment notifications are all over the place. Are we removing native_activity(), deprecating it at least, or leaving it to a separate PR entirely?

Think we can hold off for with releasing until your PR lands, I removed my changes. Got to go afk for a while, will add a CHANGELOG afterwards.

**Comment by MarijnS95** [comment]
We can do a non-breaking patch release (same for the backports) now with `#[deprecated = "Don't use for `vm()` purposes"]`, preparing everyone for this future. Or without the deprecation, as my PR would be breaking anyway.

**Comment by zarik5** [comment]
The crate oboe requires the VM and Context pointers, but also the sdk version. Is there a way of getting the sdk version from the context pointer?

**Comment by dvc94ch** [comment]
yes, should be possible, although will require fiddling a bit with the jni.

https://developer.android.com/reference/android/os/Build

**Comment by zarik5** [comment]
@dvc94ch Oh that's great, thanks!

**Pull request merged**
